### PR TITLE
point shop

### PR DESCRIPTION
**Pull request by npunati27**
Added a new database "points" with a collection for item data, "items" and quantity "quantities". 

- GET: /shop/ returns a list of all shop items 
- POST: /shop/ puts a new item into both collections
- DELETE: /shop/:ITEMID deletes an item by itemId
- PUT: /shop/quantity updates quantity of an item
- PUT: /shop/item updates item info of an item

### CODE REVIEW

File: src/database/models.ts [hunk 1]
```diff
@@ -67,31 +76,39 @@ export default class Models {
...
         this.AttendeeMetadata = getModel(AttendeeMetadata, Database.ATTENDEE, AttendeeCollection.METADATA);
         this.AttendeeProfile = getModel(AttendeeProfile, Database.ATTENDEE, AttendeeCollection.PROFILE);
+        this.AttendeeFollowing = getModel(AttendeeFollowing, Database.ATTENDEE, AttendeeCollection.FOLLOWING);
         this.AuthInfo = getModel(AuthInfo, Database.AUTH, AuthCollection.INFO);
         this.AdmissionDecision = getModel(AdmissionDecision, Database.ADMISSION, AdmissionCollection.DECISION);
         this.StaffEvent = getModel(StaffEvent, Database.EVENT, EventCollection.STAFF_EVENTS);
         this.PublicEvent = getModel(PublicEvent, Database.EVENT, EventCollection.PUBLIC_EVENTS);
         this.EventMetadata = getModel(EventMetadata, Database.EVENT, EventCollection.METADATA);
         this.EventAttendance = getModel(EventAttendance, Database.EVENT, EventCollection.ATTENDANCE);
+        this.EventFollowing = getModel(EventFollowers, Database.EVENT, EventCollection.FOLLOWERS);
```
Comment (AydanPirani) [inline src/database/models.ts]:
```suggestion
        this.EventFollowers = getModel(EventFollowers, Database.EVENT, EventCollection.FOLLOWERS);
```

and change everywhere else

File: src/services/shop/shop-router.ts [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+
+        const itemsMap = shopItems.reduce((acc: Record<string, ShopItemFormat>, item: ShopItemFormat) => {
+            acc[item.itemId] = item;
+            return acc;
+        }, {});
+
+        const quantitiesMap = shopQuantities.reduce((acc: Record<string, QuantityFormat>, quantity: QuantityFormat) => {
+            acc[quantity.itemId] = quantity;
+            return acc;
+        }, {});
```
Comment (AydanPirani) [inline src/services/shop/shop-router.ts]: Why are you reducing here? Might be better off using a map function

File: src/services/shop/shop-router.ts [hunk 3]
```diff
@@ -0,0 +1,275 @@
...
+
+        const itemIds = new Set([...Object.keys(itemsMap), ...Object.keys(quantitiesMap)]);
+
+        const itemsWithQuantity: ItemFormat[] = Array.from(itemIds).map((itemId: string) => {
+            const item: ShopItemFormat = itemsMap[itemId] || {
+                itemId,
+                name: "",
+                price: 0,
+                isRaffle: false,
+            };
```
Comment (AydanPirani) [inline src/services/shop/shop-router.ts]: Do it the other way - iterate over all items, and return quantity 0 by default if it doesn't exist

File: src/services/shop/shop-router.ts [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+            const quantity: QuantityFormat = quantitiesMap[itemId] || { itemId, quantity: 0 };
+
+            return {
+                itemId,
+                name: item.name || "",
+                price: item.price || 0,
+                isRaffle: item.isRaffle || false,
+                quantity: quantity.quantity || 0,
```
Comment (AydanPirani) [inline src/services/shop/shop-router.ts]: If id is in quantity but not in the main items db, then just ignore it/print nothing

**Review by AydanPirani — CHANGES_REQUESTED** [review]

File: src/services/shop/shop-router.ts [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+shopRouter.post("/", strongJwtVerification, async (req: Request, res: Response, next: NextFunction) => {
+    const payload: JwtPayload = res.locals.payload as JwtPayload;
+    // Check if the token has staff permissions
+    if (!hasStaffPerms(payload)) {
+        return next(new RouterError(StatusCode.ClientErrorForbidden, "InvalidPermission"));
+    }
+
+    // Convert item request into the item format
+    const itemFormatString: string = req.body;
+    const itemFormat: ItemFormat = JSON.parse(itemFormatString);
```
Comment (AydanPirani) [inline src/services/shop/shop-router.ts:118]: this doesn't look right i think

Comment (npunati27) [reply src/services/shop/shop-router.ts:118]: i tried doing:
```ts
const itemFormat: ItemFormat = req.body as ItemFormat; 
```
but i kept getting weird errors so i did this. is there another way to handle it?

Comment (AydanPirani) [reply src/services/shop/shop-router.ts:118]: hm, what weird errors are you talking about?

**Pull request closed without merging**